### PR TITLE
refactor(args): Switch to pico-args

### PR DESCRIPTION
**Pull request by Geobert**
Fixes https://github.com/rust-analyzer/rust-analyzer/issues/1768

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -930,9 +935,9 @@ dependencies = [
 name = "ra_cli"
 version = "0.1.0"
 dependencies = [
- "clap 2.33.0 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (matklad) [inline Cargo.lock]: looks like something else is still using `clap` in the crate graph?

Comment (Geobert) [reply Cargo.lock]: oops, you're totally right, it's `ra_tools`

File: crates/ra_cli/src/main.rs [hunk 2]
```diff
@@ -13,93 +14,124 @@ type Result<T> = std::result::Result<T, Box<dyn Error + Send + Sync>>;
...
-                println!("{:#?}", file.syntax());
+
+    let subcommand = std::env::args_os().nth(1);
+    if subcommand.is_none() {
+        help::print_global_help();
+        return Ok(());
+    }
+    let subcommand = subcommand.unwrap();
+    let mut args: Vec<_> = std::env::args_os().collect();
+    let mut matches = Arguments::from_vec(args.drain(2..).collect());
```
Comment (bjorn3) [inline crates/ra_cli/src/main.rs]:
```suggestion
    let mut matches = Arguments::from_vec(std::env::args_os().skip(2).collect());
```

File: crates/ra_cli/src/main.rs [hunk 3]
```diff
@@ -13,93 +14,123 @@ type Result<T> = std::result::Result<T, Box<dyn Error + Send + Sync>>;
...
+        "analysis-bench" => {
+            if matches.contains(["-h", "--help"]) {
+                help::print_analysis_bench_help();
+                return Ok(());
             } else {
-                panic!("either --highlight or --complete must be set")
-            };
-            analysis_bench::run(verbose, path.as_ref(), op)?;
+                let verbose = matches.contains(["-v", "--verbose"]);
+                let path = matches.value_from_str("--path")?.unwrap_or("".to_string());
```
Comment (matklad) [inline crates/ra_cli/src/main.rs]: unwrap_or_default

File: crates/ra_cli/src/main.rs [hunk 4]
```diff
@@ -13,93 +14,123 @@ type Result<T> = std::result::Result<T, Box<dyn Error + Send + Sync>>;
...
+        _ => help::print_global_help(),
     }
     Ok(())
 }
 
+fn handle_extra_flags(e: pico_args::Error) -> Result<()> {
+    if let pico_args::Error::UnusedArgsLeft(flags) = e {
+        let mut invalid_flags = String::new();
+        for flag in flags {
+            write!(&mut invalid_flags, "{}, ", flag).expect("Error on write");
```
Comment (matklad) [inline crates/ra_cli/src/main.rs]: let's `?`

File: crates/ra_cli/src/main.rs [hunk 5]
```diff
@@ -13,93 +14,123 @@ type Result<T> = std::result::Result<T, Box<dyn Error + Send + Sync>>;
...
-            let file = file()?;
-            if !matches.is_present("no-dump") {
-                println!("{:#?}", file.syntax());
+
+    let subcommand = std::env::args_os().nth(1);
+    if subcommand.is_none() {
+        help::print_global_help();
+        return Ok(());
+    }
+    let subcommand = subcommand.unwrap();
```
[the remark this answers is not in the document]
Comment (Geobert) [reply crates/ra_cli/src/main.rs]: nested `match` then?

File: crates/ra_cli/src/main.rs (hunk 4 above)
Comment (Geobert) [reply crates/ra_cli/src/main.rs]: You mean no need for `handle_eytra_flags`?

File: crates/ra_cli/src/main.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (matklad) [reply crates/ra_cli/src/main.rs]: match with an early return rather

File: crates/ra_cli/src/help.rs [hunk 6]
```diff
@@ -0,0 +1,96 @@
+pub fn print_global_help() {
```
[the remark this answers is not in the document]
Comment (Geobert) [reply crates/ra_cli/src/help.rs]: only `print_no_param_subcommand_help` in `ra_tools` can't be converted as it takes a parameter.

File: crates/ra_cli/src/main.rs (hunk 4 above)
Comment (matklad) [reply crates/ra_cli/src/main.rs]: I mean replace `.expect` with `?`

**Comment by matklad** [comment]
bors r+

Thanks!

**Pull request merged**